api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto

```diff
@@ -0,0 +1,105 @@
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: Payload-To-Metadata Filter]
+//
+// The configuration for transforming payloads into metadata. This is useful
+// for matching load balancer subsets, logging, etc.
+//
+// Payload to Metadata :ref:`configuration overview <config_thrift_filters_payload_to_metadata>`.
+// [#extension: envoy.filters.thrift.payload_to_metadata]
+
+message PayloadToMetadata {
+  enum ValueType {
+    STRING = 0;
+    NUMBER = 1;
+  }
+
+  // [#next-free-field: 6]
+  message KeyValuePair {
+    // The namespace — if this is empty, the filter's namespace will be used.
+    string metadata_namespace = 1;
+
+    // The key to use within the namespace.
+    string key = 2 [(validate.rules).string = {min_len: 1}];
+
+    oneof value_type {
+      // The value to pair with the given key.
+      //
+      // When used for on_present case, if value is non-empty it'll be used instead
+      // of the field value. If both are empty, the field value is used as-is.
+      //
+      // When used for on_missing case, a non-empty value must be provided.
+      string value = 3;
+
+      // If present, the header's value will be matched and substituted with this.
+      // If there is no match or substitution, the field value is used as-is.
+      //
+      // This is only used for on_present.
+      type.matcher.v3.RegexMatchAndSubstitute regex_value_rewrite = 4;
+    }
+
+    // The value's type — defaults to string.
+    ValueType type = 5 [(validate.rules).enum = {defined_only: true}];
+  }
+
+  // A Rule defines what metadata to apply when a field is present or missing.
+  // [#next-free-field: 6]
+  message Rule {
+    oneof match_specifier {
+      option (validate.required) = true;
+
+      // If specified, the route must exactly match the request method name. As a special case,
+      // an empty string matches any request method name.
+      string method_name = 1;
+
+      // If specified, the route must have the service name as the request method name prefix.
+      // As a special case, an empty string matches any service name. Only relevant when service
+      // multiplexing.
+      string service_name = 2;
+    }
+
+    // Specifies that a match will be performed on the value of a field.
+    FieldSelector field_selector = 3 [(validate.rules).message = {required: true}];
+
+    // If the field is present, apply this metadata KeyValuePair.
+    KeyValuePair on_present = 4;
+
+    // If the field is missing, apply this metadata KeyValuePair.
+    //
+    // The value in the KeyValuePair must be set, since it'll be used in lieu
+    // of the missing field value.
+    KeyValuePair on_missing = 5;
+  }
+
+  message FieldSelector {
+    // field name to log
+    string name = 1 [(validate.rules).string = {min_len: 1}];
+
+    // field id to match. Limit to the range of int16 to fit thrift format.
+    google.protobuf.Int32Value id = 2 [
```

Comment:
Can this just be an int32? It's a little easier to deal with given that it's required anyway.

Reply:
we want `int16` but if we don't set that is required for `int16`. IIRC it would be default to 0 without setting which is error-prone.

Reply:
But the validations make it required so it’s an error either way.

Reply:
Yeah this doesn't have to be Int32Value but could just be int16 since it is required anyway.

Reply:
I did a quick test and it doesn't raise `ProtoValidationException` or any other exception. 
Am i missing anything, @lizan ? But yeah i can switch to int32 if that's preferred.

```
[ RUN      ] PayloadToMetadataFilterConfigTest.SimpleConfigWithoutId
[       OK ] PayloadToMetadataFilterConfigTest.SimpleConfigWithoutId (105 ms)
``` 


```
diff --git a/api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto b/api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto
index 0148a22626..894df76e1d 100644
--- a/api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto
+++ b/api/envoy/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/v3/payload_to_metadata.proto
@@ -4,8 +4,6 @@ package envoy.extensions.filters.network.thrift_proxy.filters.payload_to_metadat
 
 import "envoy/type/matcher/v3/regex.proto";
 
-import "google/protobuf/wrappers.proto";
-
 import "udpa/annotations/status.proto";
 import "validate/validate.proto";
 
@@ -90,11 +88,8 @@ message PayloadToMetadata {
     // field name to log
     string name = 1 [(validate.rules).string = {min_len: 1}];
 
-    // field id to match. Limit to the range of int16 to fit thrift format.
-    google.protobuf.Int32Value id = 2 [
-      (validate.rules).int32 = {lte: 32767 gte: -32768},
-      (validate.rules).message = {required: true}
-    ];
+    // field id to match
+    int32 id = 2 [(validate.rules).int32 = {lte: 32767 gte: -32768}];
 
     // next node of the field selector
     FieldSelector child = 3;
diff --git a/source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc b/source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc
index c90911a572..5476eb5720 100644
--- a/source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc
+++ b/source/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter.cc
@@ -65,7 +65,7 @@ Rule::Rule(const ProtoRule& rule, uint16_t rule_id, TrieSharedPtr root)
   const FieldSelector* field_selector = &rule_.field_selector();
   TrieSharedPtr node = root;
   while (true) {
-    int16_t id = static_cast<int16_t>(field_selector->id().value());
+    int16_t id = static_cast<int16_t>(field_selector->id());
     if (node->children_.find(id) == node->children_.end()) {
       node->children_[id] = std::make_shared<Trie>(node);
     }
diff --git a/test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/config_test.cc b/test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/config_test.cc
index 8dacdf50e2..f112c21d92 100644
--- a/test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/config_test.cc
+++ b/test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/config_test.cc
@@ -183,6 +183,25 @@ void testForbiddenConfig(const std::string& yaml, const std::string& message) {
                             EnvoyException, message);
 }
 
+// does not allow rule without id in field selector
+TEST(PayloadToMetadataFilterConfigTest, SimpleConfigWithoutId) {
+  const std::string yaml = R"EOF(
+request_rules:
+  - method_name: foo
+    field_selector:
+      name: info
+    on_present:
+      metadata_namespace: envoy.lb
+      key: version
+    on_missing:
+      metadata_namespace: envoy.lb
+      key: default
+      value: unknown
+  )EOF";
+
+  testValidConfig(yaml);
+}
+
 // does not allow rule without either on_present or on_missing
 TEST(PayloadToMetadataFilterConfigTest, EmptyOnPresentOnMissing) {
   std::string yaml =

```